### PR TITLE
Change impossible match arms to not contribute to the expressions' resulting union type

### PR DESCRIPTION
**Pull request by mad-briller**
Let me know if there's something i can do better!

went down the wrong rabbit hole for hours trying to modify the NodeScopeResolver, is there any docs / resources that would be good to learn about how scopes / type inference works internally @ondrejmirtes ? something like that video of Jan talking about types you posted the other day?

### CODE REVIEW

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1534,7 +1542,9 @@ private function resolveType(Expr $node): Type
 				}
 
 				$truthyScope = $matchScope->filterByTruthyValue($filteringExpr);
-				$types[] = $truthyScope->getType($arm->body);
+				if (count($arm->conds) !== $impossibleConds) {
+					$types[] = $truthyScope->getType($arm->body);
+				}
```
Comment (rvanvelzen) [inline src/Analyser/MutatingScope.php]: An alternative would be:

```php
$filteringExprType = $matchScope->getType($filteringExpr);

if (!(new ConstantBooleanType(false))->isSuperTypeOf($filteringExprType)->yes()) {
	$truthyScope = $matchScope->filterByTruthyValue($filteringExpr);
	$types[] = $truthyScope->getType($arm->body);
}
```
... but I'm starting to wonder if `filterByTruthyValue` should be handling constant-false expressions better.

Comment (rvanvelzen) [reply src/Analyser/MutatingScope.php]: Ah, this is really unrelated to `filterByTruthyValue`. I do think my suggestion is a bit simpler though :)

Comment (mad-briller) [reply src/Analyser/MutatingScope.php]: i was trying to understand how `filterByTruthyValue` and `filterByFalseyValue` worked for a while. i think my mental model of the Scope object is lacking

a much less naive approach, thank you

**Comment by ondrejmirtes** [comment]
Thank you!

About the internals - you'll learn a lot by reading https://phpstan.org/developing-extensions/core-concepts and also maybe https://phpstan.org/writing-php-code/narrowing-types (narrowing types is implemented in TypeSpecifier class).

Other than that I'd recommend you to read release notes and see how various bugs are fixed around the codebase. Then try to fix a bug you choose :)

**Pull request merged**